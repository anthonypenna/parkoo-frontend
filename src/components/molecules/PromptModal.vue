<template>
  <Modal>
    <slot />
    <div class="prompt-modal__button-container">
      <Button @click.native="onClose">
        <slot name="close" />
      </Button>
      <div class="prompt-modal__button-separator"></div>
      <Button theme="primary" @click.native="onConfirm">
        <slot name="confirm" />
      </Button>
    </div>
  </Modal>
</template>

<script lang="ts">
import Vue from 'vue'
import Modal from '@/components/atoms/Modal.vue'
import Button from '@/components/atoms/Button.vue'

export default Vue.extend({
  name: 'PromptModal',
  components: { Modal, Button },
  methods: {
    onClose() {
      this.$emit('close')
    },
    onConfirm() {
      this.$emit('confirm')
    }
  }
})
</script>

<style lang="scss" scoped>
.prompt-modal {
  &__button-container {
    display: flex;
  }

  &__button-separator {
    margin: 0 0.5rem;
  }
}
</style>
