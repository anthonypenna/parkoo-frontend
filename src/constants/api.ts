export const API_BASEURL = process.env.VUE_APP_API_BASEURL || ''
