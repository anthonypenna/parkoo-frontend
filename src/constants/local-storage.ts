export const LAST_KNOWN_POSITION_KEY = 'lkp'
