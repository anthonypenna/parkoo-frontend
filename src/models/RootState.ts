import { DateState } from './DateState'
import { MapboxState } from './MapboxState'
import { StreetsState } from './StreetsState'
import { UserState } from './UserState'

export interface RootState {
  date: DateState
  mapbox: MapboxState
  streets: StreetsState
  user: UserState
}
