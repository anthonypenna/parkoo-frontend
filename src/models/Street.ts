export interface Street {
  id: string
  lat: number
  lng: number
  cleaningDays: Record<0 | 1 | 2 | 3 | 4 | 5 | 6, boolean>
}
