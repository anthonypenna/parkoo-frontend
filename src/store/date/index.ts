import { getters } from './getters'
import { state } from './state'

export const date = {
  namespaced: true,
  getters,
  state
}
