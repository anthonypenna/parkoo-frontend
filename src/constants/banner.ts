import { BannerState } from '@/models/BannerState'

export const BANNER_TIMEOUT = 3000

export enum BannerType {
  Success,
  Error
}

export const CLOSED_BANNER_STATE = Object.freeze<BannerState>({
  text: '',
  type: BannerType.Error,
  visible: false
})
