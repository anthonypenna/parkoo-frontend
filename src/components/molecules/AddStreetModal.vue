<template>
  <PromptModal @close="onCancel" @confirm="onAdd">
    <h2>Add new street</h2>
    <p>
      <slot name="street-name" />
    </p>
    <h3 class="add-street-modal__subsection">Cleaning days</h3>
    <div class="add-street-modal__cleaning-days">
      <div
        v-for="(day, index) of cleaningDays"
        :key="day.id"
        class="add-street-modal__cleaning-day"
      >
        <Checkbox
          :name="day.id.toString()"
          v-model="cleaningDays[index].cleaned"
        >
          <template #label>{{ day.label }}</template>
        </Checkbox>
      </div>
    </div>
    <template #close>Cancel</template>
    <template #confirm>Add</template>
  </PromptModal>
</template>

<script lang="ts">
import Vue from "vue";
import PromptModal from "@/components/molecules/PromptModal.vue";
import Checkbox from "@/components/atoms/Checkbox.vue";
import { Day } from "@/constants/date";
import { Street } from "@/models/Street";
import { mapGetters } from "vuex";

export default Vue.extend({
  name: "AddStreetModal",
  components: { PromptModal, Checkbox },
  data() {
    return {
      cleaningDays: [
        { id: Day.Monday, label: Day[Day.Monday], cleaned: false },
        { id: Day.Tuesday, label: Day[Day.Tuesday], cleaned: false },
        { id: Day.Wednesday, label: Day[Day.Wednesday], cleaned: false },
        { id: Day.Thursday, label: Day[Day.Thursday], cleaned: false },
        { id: Day.Friday, label: Day[Day.Friday], cleaned: false },
        { id: Day.Saturday, label: Day[Day.Saturday], cleaned: false },
        { id: Day.Sunday, label: Day[Day.Sunday], cleaned: false },
      ],
    };
  },
  computed: {
    ...mapGetters("user", ["lat", "lng"]),
  },
  methods: {
    onCancel() {
      this.$emit("cancel");
    },
    onAdd() {
      const street: Street = {
        id: "",
        lat: this.lat,
        lng: this.lng,
        cleaningDays: this.cleaningDays.reduce<Record<number, boolean>>(
          (days, day) => {
            days[day.id] = day.cleaned;
            return days;
          },
          {}
        ),
      };

      this.$emit("add", street);
    },
  },
});
</script>

<style lang="scss" scoped>
.add-street-modal {
  &__subsection {
    text-align: start;
  }

  &__cleaning-days {
    display: flex;
    flex-flow: column wrap;
    height: 150px;
  }

  &__cleaning-day {
    width: 50%;
  }
}
</style>
