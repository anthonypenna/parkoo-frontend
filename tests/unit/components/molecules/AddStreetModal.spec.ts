import { createLocalVue, mount, shallowMount, Wrapper } from '@vue/test-utils'
import AddStreetModal from '@/components/molecules/AddStreetModal.vue'
import PromptModal from '@/components/molecules/PromptModal.vue'
import Vuex from 'vuex'
import { RootState } from '@/models/RootState'
import { Street } from '@/models/Street'
import { Day } from '@/constants/date'

type AddStreetModalComponent = Wrapper<
  Vue & {
    cleaningDays: Array<{ id: Day; label: string; cleaned: boolean }>
  }
>

type PromptModalComponent = Wrapper<
  Vue & {
    onClose(): void
    onConfirm(): void
  }
>

const localVue = createLocalVue()

localVue.use(Vuex)

const store = new Vuex.Store<RootState>({
  getters: {
    nameOfStreetBeingAdded: jest.fn()
  },
  modules: {
    user: {
      namespaced: true,
      state: {
        lat: 0,
        lng: 0
      },
      getters: {
        lat: jest.fn(() => 0),
        lng: jest.fn(() => 0)
      }
    }
  }
})

describe('<AddStreetModal />', () => {
  it('should render correctly', () => {
    const wrapper = shallowMount(AddStreetModal, { localVue, store })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when clicking on the cancel button', () => {
    it('should emit a cancel event', () => {
      const wrapper = mount(AddStreetModal, {
        localVue,
        store,
        stubs: { Modal: true }
      })

      const modal = wrapper.findComponent(PromptModal) as PromptModalComponent
      modal.vm.onClose()
      expect(wrapper.emitted().cancel).toBeTruthy()
    })
  })

  describe('when clicking on the add button', () => {
    it('should emit an add event', () => {
      const wrapper = mount(AddStreetModal, {
        localVue,
        store,
        stubs: { Modal: true }
      })

      const modal = wrapper.findComponent(PromptModal) as PromptModalComponent
      modal.vm.onConfirm()
      expect(wrapper.emitted().add).toBeTruthy()
    })

    it('should emit the correct payload', () => {
      const wrapper = mount(AddStreetModal, {
        localVue,
        store,
        stubs: { Modal: true }
      }) as AddStreetModalComponent

      wrapper.vm.cleaningDays[0].cleaned = true

      const modal = wrapper.findComponent(PromptModal) as PromptModalComponent
      modal.vm.onConfirm()

      const addEvents = wrapper.emitted().add as unknown[][]
      expect(addEvents[0][0]).toEqual<Street>({
        cleaningDays: {
          0: false,
          1: true,
          2: false,
          3: false,
          4: false,
          5: false,
          6: false
        },
        id: '',
        lat: 0,
        lng: 0
      })
    })
  })
})
