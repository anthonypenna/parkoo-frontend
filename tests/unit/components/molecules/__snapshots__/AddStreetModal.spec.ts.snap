// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddStreetModal /> should render correctly 1`] = `
<promptmodal-stub>
  <h2>Add new street</h2>
  <p></p>
  <h3 class="add-street-modal__subsection">Cleaning days</h3>
  <div class="add-street-modal__cleaning-days">
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="1"></checkbox-stub>
    </div>
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="2"></checkbox-stub>
    </div>
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="3"></checkbox-stub>
    </div>
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="4"></checkbox-stub>
    </div>
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="5"></checkbox-stub>
    </div>
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="6"></checkbox-stub>
    </div>
    <div class="add-street-modal__cleaning-day">
      <checkbox-stub name="0"></checkbox-stub>
    </div>
  </div>
</promptmodal-stub>
`;
